src/daft-json/src/local.rs

```diff
@@ -0,0 +1,485 @@
+            .as_ref()
+            .and_then(|options| options.predicate.clone());
+
+        let n_threads: usize = std::thread::available_parallelism()
+            .unwrap_or(NonZeroUsize::new(2).unwrap())
+            .into();
+
+        let schema = match convert_options
+            .as_ref()
+            .and_then(|options| options.schema.as_ref())
+        {
+            Some(schema) => schema.as_ref().clone(),
+            None => Schema::try_from(&infer_schema(bytes, None, None)?)?,
+        };
+
+        let pool = if let Some(max_in_flight) = max_chunks_in_flight {
+            ThreadPoolBuilder::new().num_threads(max_in_flight).build()
+        } else {
+            ThreadPoolBuilder::new().num_threads(n_threads).build()
+        }
+        .context(RayonThreadPoolSnafu)?;
+
+        let projected_schema = match convert_options.and_then(|options| options.include_columns) {
+            Some(projected_columns) => schema.project(&projected_columns)?,
+            None => schema,
+        };
+
+        Ok(Self {
+            bytes,
+            schema: projected_schema,
+            predicate,
+            n_threads,
+            sample_size,
+            n_rows,
+            chunk_size,
+            pool,
+        })
+    }
+
+    pub fn finish(&self) -> DaftResult<Table> {
+        let mut bytes = self.bytes;
+        let mut n_threads = self.n_threads;
+        let mut total_rows = 128;
+
+        if let Some((mean, std)) = get_line_stats_json(bytes, self.sample_size) {
+            let line_length_upper_bound = mean + 1.1 * std;
+
+            total_rows = (bytes.len() as f32 / (mean - 0.01 * std)) as usize;
+            if let Some(n_rows) = self.n_rows {
+                total_rows = std::cmp::min(n_rows, total_rows);
+                // the guessed upper bound of the no. of bytes in the file
+                let n_bytes = (line_length_upper_bound * (n_rows as f32)) as usize;
+
+                if n_bytes < bytes.len() {
+                    if let Some(pos) = next_line_position(&bytes[n_bytes..]) {
+                        bytes = &bytes[..n_bytes + pos]
+                    }
+                }
+            }
+        }
+
+        if total_rows <= 128 {
+            n_threads = 1;
+        }
+
+        let total_len = bytes.len();
+        let chunk_size = self.chunk_size.unwrap_or_else(|| total_len / n_threads);
+        let file_chunks = self.get_file_chunks(bytes, n_threads, total_len, chunk_size);
+
+        let tbls = self.pool.install(|| {
+            file_chunks
+                .into_par_iter()
+                .map(|(start, stop)| {
+                    let chunk = &bytes[start..stop];
+                    self.parse_json_chunk(chunk, chunk_size)
+                })
+                .collect::<DaftResult<Vec<Table>>>()
+        })?;
+
+        let tbl = tables_concat(tbls)?;
+
+        // The `limit` is not guaranteed, so we need to properly apply the limit after concatenating the
```

Comment:
Unfinished comment.

Reply:
We could probably apply the limit pre-concatenation, i.e. eliding concatenation of tables outside of the limit, right? Although we can definitely do that as a follow-up (we often do a naive post-concat limit application elsewhere too).

Reply:
So we do still partially apply the limit pre-concat. If `n_rows` is applied, we take a smaller slice of `bytes`, leading to an approximate `n_rows` before we do any parsing. The limit after is to guarantee the exact number of rows.